envoy: Better track updates when qualifying Envoy resource names

jrajahalme opened the pull request:
Pass a boolean "updated" pointer to api.ResourceQualifiedName() in order to better track if updates were done or not.

In the case of Http Connection Manager RouteConfig the need for the update was ignored, so this is a bug fix for that.

Factor out the TCP proxy qualification logic to
qualifyTcpProxyResourceNames() which is similar to qualifyRouteConfigurationResourceNames() to keep the main logic simpler.

Fixes: #26037

```release-note
"envoy-admin" cluster is renamed as "/envoy-admin", requiring all references in CEC/CCEC to be updated.
```

mhofstetter commented on pkg/policy/api/utils.go:
```diff
@@ -146,7 +146,7 @@ const (
 )
 
-func ResourceQualifiedName(namespace, cecName, resourceName string, options ...Option) string {
+func ResourceQualifiedName(namespace, cecName, resourceName string, updated *bool, options ...Option) string {
```
Personally i'd prefer having the updated `bool` as secondary return parameter instead of passing it as pointer here too (like in the using functions). Even if it might makes the usage of it a little bit more complex (assignment & check that we don't override a updated `true` with `false in cases were we execute the function multiple times)

jrajahalme replied:
I considered it initially, but it gets quote cumbersome due to assignment to a temporary and then comparing that we do not overwrite a `true` with a `false`.

I see the inconsistency, though. How about changing the "using" functions to take a pointer as well?

mhofstetter replied:
Actually it's more about the "pattern" ("unexpected" side-effect on an input parameter) than the inconsistency with the other functions. IMO, it's even better to only have it this way in the inner function.

pippolo84 replied:
I second Marco's concern here. This pattern forces us to pass a `nil` when we are not interested in `updated`, making the helper more cumbersome. Also, it requires us to check every reference to `updated` in `ResourceQualifiedName` body to not be `nil`.

mhofstetter reviewed (approved):
Thanks for fixing the issue! I won't block this PR for a personal opinion on the "pointer passing" part! :smiley:

pippolo84 commented on pkg/envoy/ciliumenvoyconfig.go:
```diff
@@ -81,30 +81,45 @@ func (old *Resources) ListenersAddedOrDeleted(new *Resources) bool {
 }
 
-func qualifyRouteConfigurationResourceNames(namespace, name string, routeConfig *envoy_config_route.RouteConfiguration) {
+func qualifyTcpProxyResourceNames(namespace, name string, tcpConfig *envoy_config_tcp.TcpProxy) (updated bool) {
```
Nit: maybe moving the introduction of this helper in a separate commit makes things a bit clearer.

jrajahalme replied:
Done.

pippolo84 reviewed (approved):
Thanks for the fix! :+1: 
I've left a couple of comments inline, but I understand that, being also a bugfix, we might not want to block this further. Maybe we can improve the `updated` pattern in a follow up PR?

jrajahalme committed:
envoy: Add qualifyTcpProxyResourceNames helper

Factor out TcpProxy resource name qualifications to a helper function and
check for nil pointers.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Better track updates when qualifying Envoy resource names

Pass a boolean "updated" pointer to api.ResourceQualifiedName() in order
to better track if updates were done or not.

In the case of Http Connection Manager RouteConfig the need for the
update was ignored, so this is a bug fix for that. This omission hid a
bug where HCM RouteConfigs in our examples (and tests) refer to
"envoy-admin" cluster while unqualified cluster names should no longer be
accessible to CEC/CCEC. Fix this by renaming "envoy-admin" as
"/envoy-admin". This requres all references to "envoy-admin" cluster to
be changed to "/envoy-admin", including in any CEC/CCEC users may have.

Factor out the TCP proxy qualification logic to
qualifyTcpProxyResourceNames() which is similar to
qualifyRouteConfigurationResourceNames() to keep the main logic simpler.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
